pkg/apiserver/apic.go

```diff
@@ -0,0 +1,231 @@
+	metricsInterval time.Duration
+	dbClient        *database.Client
+	apiClient       *apiclient.ApiClient
+	alertToPush     chan []*models.Alert
+	mu              sync.Mutex
+	pushTomb        tomb.Tomb
+	pullTomb        tomb.Tomb
+	metricsTomb     tomb.Tomb
+	startup         bool
+	credentials     *csconfig.ApiCredentialsCfg
+}
+
+func NewAPIC(config *csconfig.OnlineApiClientCfg, dbClient *database.Client) (*apic, error) {
+	var err error
+	var ret *apic
+
+	apiclient.BaseURL, err = url.Parse(config.Credentials.URL)
+	if err != nil {
+		return ret, errors.Wrapf(err, "parse local API URL '%s': %v ", config.Credentials.URL, err.Error())
+	}
+
+	password := strfmt.Password(config.Credentials.Password)
+	t := &apiclient.JWTTransport{
+		MachineID: &config.Credentials.Login,
+		Password:  &password,
+	}
+
+	pullInterval, err := time.ParseDuration(PullInterval)
+	if err != nil {
+		return ret, err
+	}
+	pushInterval, err := time.ParseDuration(PushInterval)
+	if err != nil {
+		return ret, err
+	}
+	metricsInterval, err := time.ParseDuration(MetricsInterval)
+	if err != nil {
+		return ret, err
+	}
+
+	return &apic{
+		apiClient:       apiclient.NewClient(t.Client()),
+		alertToPush:     make(chan []*models.Alert),
+		dbClient:        dbClient,
+		pullInterval:    pullInterval,
+		pushInterval:    pushInterval,
+		metricsInterval: metricsInterval,
+		mu:              sync.Mutex{},
+		startup:         true,
+		credentials:     config.Credentials,
+		pullTomb:        tomb.Tomb{},
+		pushTomb:        tomb.Tomb{},
+		metricsTomb:     tomb.Tomb{},
+	}, nil
+}
+
+func (a *apic) Push() error {
+	defer types.CatchPanic("apil/pushToAPIC")
+
+	var cache []*models.Alert
+	ticker := time.NewTicker(a.pushInterval)
+
+	for {
+		select {
+		case <-a.pushTomb.Dying(): // if one apic routine is dying, do we kill the others?
+			a.pullTomb.Kill(nil)
+			a.metricsTomb.Kill(nil)
+			log.Infof("push tomb is dying, sending cache (%d elements) before exiting", len(cache))
+			err := a.Send(cache)
+			return err
+		case <-ticker.C:
+			// flush
+			a.mu.Lock()
+			err := a.Send(cache)
+			if err != nil {
+				return err
+			}
+			cache = make([]*models.Alert, 0)
+			a.mu.Unlock()
+		case alerts := <-a.alertToPush:
+			a.mu.Lock()
+			cache = append(cache, alerts...)
+			a.mu.Unlock()
+		}
+	}
+}
+
+func (a *apic) Send(cache []*models.Alert) error {
+	_, _, err := a.apiClient.Alerts.Add(context.Background(), cache)
+	return err
+}
+
+func (a *apic) Pull() error {
+	defer types.CatchPanic("apil/pullFromAPIC")
+
+	ticker := time.NewTicker(a.pullInterval)
+	for {
+		select {
+		case <-ticker.C:
+			data, _, err := a.apiClient.Decisions.GetStream(context.Background(), a.startup)
+			if err != nil {
+				return errors.Wrap(err, "pull top")
+			}
+			if a.startup {
+				a.startup = false
+			}
+			for _, decision := range data.New {
+				alertCreated, err := a.dbClient.Ent.Alert.
+					Create().
+					SetScenario(*decision.Scenario).
+					SetSourceIp(*decision.Value).
```

Comment:
I thin we want to set origin as consensus or something like that? It will allow user to easily see the decisions that are coming from the consensus